pxd_fastpath.c

```diff
@@ -995,19 +1051,32 @@ void pxd_suspend_io(struct pxd_device *pxd_dev)
 	}
 }
 
-// external request to resume IO on fastpath device
-int pxd_request_resume(struct pxd_device *pxd_dev)
+int pxd_request_resume_internal(struct pxd_device *pxd_dev)
 {
-	if (pxd_dev->using_blkque || !pxd_dev->fp.app_suspend) {
+	if (pxd_dev->using_blkque) {
 		return -EINVAL;
 	}
 
 	pxd_resume_io(pxd_dev);
-	pxd_dev->fp.app_suspend = false;
 	printk(KERN_NOTICE"device %llu resumed IO from userspace\n", pxd_dev->dev_id);
 	return 0;
 }
 
+// external request to resume IO on fastpath device
+int pxd_request_resume(struct pxd_device *pxd_dev)
+{
+	int rc;
+	if (atomic_read(&pxd_dev->fp.app_suspend) == 0) {
```

Comment:
Don't u need a CAS here as well? or is it ok to let parallel calls to pxd_request_resume_internal through?

Reply:
pxd_request_{resume|suspend} is userspace suspend/resume op. Only once it can be called to set/clear state.
They are blocked for multiple calls within px-storage by device state. 
But internal to px-fuse suspend/resume can be called multiple times recursively. px-fuse does not call pxd_request_xxx calls.

Reply:
Ok, but pxd_request_suspend() uses a cas.

Reply:
I see, will revise for keeping it similar.

Reply:
I made both userspace request/suspend to be same. cas is not needed.